### PR TITLE
Add file upload element to interactive dialogs

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary
  Adds a new `file` element type to interactive dialogs, allowing users to
  upload files directly within a dialog. Uploaded file IDs are validated
  server-side (existence + ownership) and forwarded to the integration on
  submit. Supports single and multiple-file fields.

  ## Changes

  **Server**
  - `model`: new `file` `DialogElement` type with validation, `AllowMultiple`
    field, and `SubmitDialogRequest.FileIds`.
  - `SubmitInteractiveDialog` validates submitted file IDs (existence +
    `CreatorId` ownership) from both `file_ids` and any file IDs referenced
    in `submission` values, batched and bounded.
  - `Client4.GetFileInfo` / `client4.getFileInfo`.

  **Webapp**
  - New `AppsFormFileUpload` component: upload, progress, removal, and
    hydration of pre-set file IDs; single vs `allow_multiple` selection.
  - Wired into the dialog → apps-form conversion as the `file` field type.
  - Submit is blocked while any field has an upload in progress (per-field
    pending tracking, so concurrent file fields don't clobber each other).

  ## Testing Done
  - Go unit tests: model validation + submit-time file-ID validation
    (ownership, dedup/limit, submission-value smuggling, non-file IDs).
  - Jest: `AppsFormFileUpload` component.
  - Cypress e2e: `file_upload_spec.js` (render, single+multi upload & submit,
    append/replace, required validation) + webhook fixtures.
  - `make check-style`, `npm run check-types`, `npm run fix` clean.

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-66109

#### Related PRs
Documentation Updates - https://github.com/mattermost/mattermost-developer-documentation/pull/1509

#### Screenshots
Screenshot 2026-03-24 at 2 46 55 PM

#### Release Note
```release-note
 Added support for a file upload element type in interactive dialogs.
```

## Change Impact: 🟡 Medium

**Regression Risk:** Changes modify server-side dialog submission/validation and the webapp file-upload UI and submit flow, touching API payloads and shared dialog conversion utilities; comprehensive tests exist but integration/hydration/ownership edge cases could still cause regressions.  
**QA Recommendation:** Moderate manual QA recommended—focus on dialog submit flows (single vs multiple files), hydrated default IDs, ownership/permission failures, submission-value smuggling, and end-to-end webhook integrations in addition to relying on the provided unit/e2e tests.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: server/channels/app/integration_action.go [hunk 1]
```diff
@@ -650,6 +748,7 @@ func (a *App) SubmitInteractiveDialog(rctx request.CTX, request model.SubmitDial
 
 	var response model.SubmitDialogResponse
 	if len(body) == 0 {
+		rctx.Logger().Info("SubmitDialogResponse empty body")
```
Comment (wiggin77) [inline server/channels/app/integration_action.go]: Is this log line meant to be here, at `Info` level?  At minimum it should have more context to make it useful.

File: server/channels/app/integration_action.go [hunk 2]
```diff
@@ -615,6 +615,102 @@ func (a *App) SubmitInteractiveDialog(rctx request.CTX, request model.SubmitDial
...
+	// Validate submitted file IDs exist and belong to the submitting user.
+	// Dedup without sorting so the order the user selected the files is preserved
+	// when the request is forwarded to the integration.
+	request.FileIds = model.RemoveDuplicateStringsNonSort(request.FileIds)
+	if len(request.FileIds) > model.MaxDialogFileIds {
+		return nil, model.NewAppError("SubmitInteractiveDialog", "app.submit_interactive_dialog.too_many_file_ids",
+			map[string]any{"Max": model.MaxDialogFileIds}, "", http.StatusBadRequest)
+	}
+	declaredFileIDs := make(map[string]bool, len(request.FileIds))
+	for _, fileID := range request.FileIds {
```
Comment (wiggin77) [inline server/channels/app/integration_action.go:335]: Below this is done with a single DB call:
```go
a.Srv().Store().FileInfo().GetByIds(candidateFileIDs, false, false, false)
```

I think we can do that here and reduce the DB roundtrips.

File: server/channels/app/integration_action.go (hunk 1 above)
Comment (sbishel) [reply server/channels/app/integration_action.go]: Thanks, for my own testing during development. I need to start using Debug for my own testing.

**Comment by sbishel** [comment]
Doug, I also put you on to review the developer documentation for this feature. Thanks.
https://github.com/mattermost/mattermost-developer-documentation/pull/1509

File: server/channels/app/integration_action.go [hunk 3]
```diff
@@ -615,6 +615,116 @@ func (a *App) SubmitInteractiveDialog(rctx request.CTX, request model.SubmitDial
...
+		switch typed := v.(type) {
+		case string:
+			for tok := range strings.SplitSeq(typed, ",") {
+				tok = strings.TrimSpace(tok)
+				if tok == "" || declaredFileIDs[tok] || seenCandidate[tok] || !model.IsValidId(tok) {
+					continue
+				}
+				seenCandidate[tok] = true
+				candidateFileIDs = append(candidateFileIDs, tok)
+				if len(candidateFileIDs) >= maxSubmissionFileIDScan {
```
Comment (wiggin77) [inline server/channels/app/integration_action.go]: If I'm reading this right, it stops collecting after it gets 256 IDs and returns. It counts invalid and valid file ids.  256 bogus ids followed by a valid file id for another user wouldn't check the last id.  

I'm not clear what happens for the caller, except that they expected all the ids to be checked but it silently didn't check all of them.

Comment (wiggin77) [reply server/channels/app/integration_action.go]: I've not found any other issues, and I'm not sure if this is a blocker or not.  I'll approve the PR assuming this will be checked.

**Review by wiggin77 — APPROVED** [review]

File: server/channels/app/integration_action.go (hunk 3 above)
Comment (wiggin77) [reply server/channels/app/integration_action.go]: @coderabbitai is this an issue?

**Review by ogi-m — APPROVED** [review]
LGTM!

Tested the following:

End-to-end

- [x] Dialog opens via OpenDialog (HTTP 200) from a slash command with a file element
- [x] Single file upload → submit → integration receives submission.single_file + aggregated file_ids
- [x] Multiple file upload → integration receives submission.multi_files (comma-joined) + file_ids aggregating all fields
- [x] Owner-validated file IDs correctly forwarded to the integration

Server-side validation (via submit API)

- [x] Own file → accepted, forwarded
- [x] Another user's file in file_ids → 403 file_not_owned
- [x] Another user's file smuggled in a submission value → 403 file_not_owned
- [x] Non-existent / non-file ID → 400 invalid_file_id
- [x] More than 10 files → 400 too_many_file_ids


Webapp UI

- [x] File field(s) render (single + allow_multiple)
- [x] Upload with progress; single replaces, multi appends
- [x] Remove an uploaded file
- [x] Required empty file field → inline validation error, submit blocked
- [x] Submit blocked while an upload is in progress
